File: src/common/OpenTelemetry.cs
using Azure.Monitor.OpenTelemetry.AspNetCore;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.DependencyInjection.Extensions;
using Microsoft.Extensions.Logging;
using OpenTelemetry.Logs;
using OpenTelemetry.Metrics;
using OpenTelemetry.Trace;
using System.Diagnostics;

namespace common;

public static class OpenTelemetryServices
{
    public static void Configure(IServiceCollection services, string activitySourceName)
    {
#pragma warning disable CA2000 // Dispose objects before losing scope
        services.TryAddSingleton(new ActivitySource(activitySourceName));
#pragma warning restore CA2000 // Dispose objects before losing scope

        services.AddOpenTelemetry()
                .WithMetrics(metrics => metrics.AddAspNetCoreInstrumentation()
                                               .AddHttpClientInstrumentation()
                                               .AddRuntimeInstrumentation()
                                               .AddMeter("*"))
                .WithTracing(tracing => tracing.AddAspNetCoreInstrumentation()
                                               .AddHttpClientInstrumentation()
                                               .AddSource("*")
                                               .SetSampler<AlwaysOnSampler>());

        var configuration = services.BuildServiceProvider().GetRequiredService<IConfiguration>();
        configuration.TryGetValue("OTEL_EXPORTER_OTLP_ENDPOINT")
                     .Iter(_ =>
                     {
                         services.AddLogging(builder => builder.AddOpenTelemetry());
                         services.Configure<OpenTelemetryLoggerOptions>(logging => logging.AddOtlpExporter())
                                 .ConfigureOpenTelemetryMeterProvider(metrics => metrics.AddOtlpExporter())
                                 .ConfigureOpenTelemetryTracerProvider(tracing => tracing.AddOtlpExporter());
                     });

        configuration.TryGetValue("APPLICATIONINSIGHTS_CONNECTION_STRING")
                     .Iter(_ => services.AddOpenTelemetry()
                                        .UseAzureMonitor());
    }
}

File: src/common/OpenTelemetry.fs
[<RequireQualifiedAccess>]
module common.OpenTelemetry

open Azure.Monitor.OpenTelemetry.AspNetCore
open Microsoft.Extensions.Configuration
open Microsoft.Extensions.DependencyInjection
open Microsoft.Extensions.DependencyInjection.Extensions
open Microsoft.Extensions.Logging
open OpenTelemetry.Logs
open OpenTelemetry.Metrics
open OpenTelemetry.Trace
open System.Diagnostics

let configureServices (services: IServiceCollection) activitySourceName =
    services.TryAddSingleton(new ActivitySource(activitySourceName))

    let _ =
        services
            .AddOpenTelemetry()
            .WithMetrics(fun metrics ->
                metrics
                    .AddAspNetCoreInstrumentation()
                    .AddHttpClientInstrumentation()
                    .AddRuntimeInstrumentation()
                    .AddMeter("*")
                |> ignore)
            .WithTracing(fun tracing ->
                tracing
                    .AddAspNetCoreInstrumentation()
                    .AddHttpClientInstrumentation()
                    .AddSource("*")
                    .SetSampler<AlwaysOnSampler>()
                |> ignore)

    let configuration = services.BuildServiceProvider().GetService<IConfiguration>()

    Configuration.tryGetValue configuration "OTEL_EXPORTER_OTLP_ENDPOINT"
    |> Option.iter (fun _ ->
        services
            .AddLogging(fun builder -> builder.AddOpenTelemetry() |> ignore)
            .Configure<OpenTelemetryLoggerOptions>(fun (options: OpenTelemetryLoggerOptions) ->
                options.AddOtlpExporter() |> ignore)
            .ConfigureOpenTelemetryMeterProvider(fun metrics -> metrics.AddOtlpExporter() |> ignore)
            .ConfigureOpenTelemetryTracerProvider(fun tracing -> tracing.AddOtlpExporter() |> ignore)
        |> ignore)

    Configuration.tryGetValue configuration "APPLICATIONINSIGHTS_CONNECTION_STRING"
    |> Option.iter (fun _ -> services.AddOpenTelemetry().UseAzureMonitor() |> ignore)
